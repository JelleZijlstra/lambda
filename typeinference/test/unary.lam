data nat = Zero | Succ nat

let rec of_integer = \n. match n with {
	0 -> Zero
	| _ when n > 0 -> Succ (of_integer (n - 1))
}

let rec to_integer = \n. match n with {
	Zero -> 0
	| Succ n' -> 1 + to_integer n'
}

let rec plus = \n1. \n2. match n1 with {
	Zero -> n2
	| Succ n1' -> plus n1' (Succ n2)
}

let rec mult = \n1. \n2. match n1 with {
	Zero -> Zero
	| Succ n1' -> plus n2 (mult n1' n2)
}

let rec exp = \n1. \n2. match n2 with {
	Zero -> Succ Zero
	| Succ n2' -> mult n1 (exp n1 n2')
}

;;

print (to_integer (exp (Succ (Succ Zero)) (Succ (Succ (Succ Zero)))))
