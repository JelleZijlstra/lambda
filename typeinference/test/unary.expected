%%stdout%%
8
Result: module
	let to_integer = \n. match n with Zero -> 0 | Succ n' -> 1 + to_integer n'
	let plus = \n1. \n2 : 'typevar13. match n1 with Zero -> n2 | Succ n1' -> plus n1' (Succ n2)
	let of_integer = \n. match n with 0 -> Zero | _ when n > 0 -> Succ (of_integer n - 1)
	let mult = \n1. \n2 : 'typevar22. match n1 with Zero -> Zero | Succ n1' -> plus n2 (mult n1' n2)
	let exp = \n1. \n2 : 'typevar33. match n2 with Zero -> Succ Zero | Succ n2' -> mult n1 (exp n1 n2')
	let __result__ = ()
end
%%stderr%%
